codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -496,19 +520,49 @@ private void incrementPendingBytes(int numBytes) {
              */
             int write(int allowedBytes) {
                 int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
+                int writtenBytes = 0;
                 try {
-                    connectionState().incrementStreamWindow(-writtenBytes);
-                    incrementStreamWindow(-writtenBytes);
+                    boolean writeError = true;
+                    try {
+                        writing = true;
+                        needFlush |= payload.write(Math.max(0, allowedBytes));
+                        writing = false;
+                        writeError = false;
+                    } catch (Throwable t) {
+                        writing = false;
+                        pendingWriteQueue.remove();
+                        payload.error(t);
+                    }
+                    if (!writeError) {
```

Comment:
@nmittler 
I think we shouldn't call cancel() directly but call FlowControlled.error() instead, which kinda makes the additional flag (it's only a local variable) necessary. Also in the above code if writeComplete throws an exception it will not be rethrown - which it should though.